### PR TITLE
Rule: Child elements of list(s) follow context model (c6f8a9)

### PR DESCRIPTION
**Pull request by jeeyyy**
A rule proposed for year 3. Feedback & reviews welcome.

Closes issue(s):
- NA

Need for Final Call:
- Yes (2 Weeks)

### CODE REVIEW

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+
+## Applicability
+
+The rule applies to every `li`, `dd` and `dt` HTML elements where the [semantic role][] is the same as the [implicit semantic role][].
+
+## Expectation
+
+Each target element:
+
+- is a [palpable content][], and;
```
Comment (Jym77) [inline _rules/lists-children-follow-context-model-c6f8a9.md:28]: `li` elements are not palpable content and will always fail this rule.

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+rule_type: atomic
+description: |
+  This rule checks that child elements of a list follow context model
+accessibility_requirements:
+  wcag20:1.3.1: # Info and Relationships (A)
+    forConformance: true
+    failed: not satisfied
+    passed: further testing needed
+    inapplicable: further testing needed
+input_aspects:
```
Comment (Jym77) [inline _rules/lists-children-follow-context-model-c6f8a9.md:13]: Need accessibility tree for the semantic role.

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+The rule applies to every `li`, `dd` and `dt` HTML elements where the [semantic role][] is the same as the [implicit semantic role][].
+
+## Expectation
+
+Each target element:
+
+- is a [palpable content][], and;
+- is contained by a node which follows the [flow content model][], and;
+- has an [explicit semantic role][] or [implicit semantic role][] that that is an allowed role by the owner element.
```
Comment (Jym77) [inline _rules/lists-children-follow-context-model-c6f8a9.md:30]: This still has the target element as subject. By applicability, the semantic role is the implicit role.

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+## Applicability
+
+The rule applies to every `li`, `dd` and `dt` HTML elements where the [semantic role][] is the same as the [implicit semantic role][].
+
+## Expectation
+
+Each target element:
+
+- is a [palpable content][], and;
+- is contained by a node which follows the [flow content model][], and;
```
Comment (Jym77) [inline _rules/lists-children-follow-context-model-c6f8a9.md:29]: "contained" (DOM parent/child relation) or "owned" (ARIA relation)? Given that the next item uses "owner element" it is not clear.

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+
+```html
+<ul role="menu">
+	<li>Milk</li>
+</ul>
+```
+
+#### Failed Example 2
+
+This `dt` element is not owned by an element that follows the [flow content model][].
```
Comment (Jym77) [inline _rules/lists-children-follow-context-model-c6f8a9.md:96]: I don't understand. `ol` is flow content.

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+
+```html
+<ol>
+	<dt>Physics</dt>
+</ol>
+```
+
```
Comment (ShadowBB) [inline _rules/lists-children-follow-context-model-c6f8a9.md:103]: A Failed example where the target element is not "a [palpable content]" is missing.

**Comment by jkodu** [comment]
Thanks @ShadowBB and @Jym77 for the reviews. I have made some change. Please review again.

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+```html
+<label>
+	<li>Milk</li>
+</label>
+```
+
+#### Failed Example 2
+
+This `dt` element is [owned by][] an element that follows the [flow content][] model, but has an [explicit semantic role][] that is not in the [required owned element][] list.
```
Comment (carlosapaduarte) [inline _rules/lists-children-follow-context-model-c6f8a9.md:97]:
```suggestion
This `dt` element is [owned by][] an element that follows the [flow content][] model, but has an [explicit semantic role][] that is not in the [required context role][] list of the `dt` element.
```

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+<dl role="columnheader">
+	<dt>Physics</dt>
+</dl>
+```
+
+### Inapplicable
+
+#### Inapplicable Example 1
+
+This rule is not applicable to `span` elements, although it is a [palpable content][].
```
Comment (carlosapaduarte) [inline _rules/lists-children-follow-context-model-c6f8a9.md:109]: Being palpable is not an applicability requirement

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+The rule applies to every `li`, `dd` and `dt` HTML elements where the [semantic role][] is the same as the [implicit semantic role][].
+
+## Expectation
+
+Each test target element:
+
+- is [owned by][] a node which;
+  - is categorized as [flow content][]; and
+  - has a [semantic role][] from the [required owned element][] list for the test target's [semantic role]().
```
Comment (Jym77) [inline _rules/lists-children-follow-context-model-c6f8a9.md:31]: This has nothing to do with the HTML content model.
Not a problem per se (still a valid thing to check), but he rule name and description are misleading…

The symmetric relation of the HTML content model is the allowed context (https://html.spec.whatwg.org/multipage/dom.html#concept-element-contexts) In the case of, say, `li` (https://html.spec.whatwg.org/multipage/grouping-content.html#the-li-element) it is indeed `ol`, `ul` and `menu`.

Is the HTML allowed context/content model what you want to check? I am not sure this has a fail-on-fail with 1.3.1 (rather with 4.1.1). Is the ARIA role model what you want to check?

File: _rules/lists-children-follow-context-model-c6f8a9.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+<ul>
+	<li>Complete online form</li>
+</ul>
+```
+
+### Failed
+
+#### Failed Example 1
+
+This `li` element is not [owned by][] an element that follows the [flow content][] model.
```
[the remark this answers is not in the document]
Comment (Jym77) [reply _rules/lists-children-follow-context-model-c6f8a9.md:87]: This fails both the HTML content model (`li` cannot be child of `label`) (which the rule does not check despite its name) and the ARIA role model (`listitem` cannot be owned by no role (`label` has no role).

**Comment by WilcoFiers** [comment]
This PR is no longer being worked on.

**Pull request closed without merging**